py/server/deephaven/table.py

```diff
@@ -503,6 +503,36 @@ def meta_table(self) -> Table:
     def j_object(self) -> jpy.JType:
         return self.j_table
 
+    def iter_rows(self, cols: Optional[Union[str, Sequence[str]]] = None, *, chunk_size: Optional[int] = 1) \
+            -> Generator[Dict[str, Any], None, None]:
```

Comment:
I wonder if we should separate the "rows" and the "chunks" methods, as we do internally. I think the caveats around `chunk_size` controlling the type of output is a bit strange (even though it's "okay" with python).

I think it would also be very nice to consider Tuple-based versions of this, as I think it's a bit stricter, safer, and explicit. What I mean by that is in the common case, the user needs to declare variables for every item returned if they are destructuring it:

```python
for foo, bar, baz in iter_rows_tuple(...):
  ...
```

The nice thing about this is that they'll get an error if they are either trying to consume too many columns or too little:

```
>>> (foo, bar, baz) = (1, 2, 3)

>>> (foo, bar, baz) = (1, 2, 3, 4)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: too many values to unpack (expected 3)

>>> (foo, bar, baz) = (1, 2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: not enough values to unpack (expected 3, got 2)

```

This also forces the caller to be more explicit; and they will get lint warnings if they have an unused variables.

I could forsee a lot of use cases where callers are lazy with the current interface and only want to get out a single column from a wide table, causing a lot of extra data to make it into the dicts that they will never use.

```python
# as written
for dict in iter_rows(...):
  foo = dict['Foo']
  # maybe the user has 20 columns that they've implicitly pulled in, but never needed
  ...
```

```python
# tuple version 
for foo in iter_rows_tuple(['Foo']):
  # the user has pulled in exactly 1 column b/c they are forced to be more explicit (otherwise they'll get an error)
   ...
```

Reply:
To be clear: I'm not arguing we create _another_ method called `iter_rows_tuple`; I'm arguing `iter_rows` should be `Tuple` based (`iter_rows_tuple` named differently just for clarity about semantics above)

Reply:
I agree with splitting the method into separate  chunk-based and row-based methods. I am not entirely sold on the `tuple` over `dict` return.  Pandas has a `iterrows` method which returns a Series indexed by the column names. It also has a `itertuples` method which returns a named tuple. The named part or the key of dict is important IMO as it forces the user to be explicit about the specific columns to read on. I considered offering the same but was concerned about the naming part of namedtuple and that it being different from the table listener interface.

Reply:
> The named part or the key of dict is important IMO as it forces the user to be explicit about the specific columns to read on.

Not sure what you mean; IMO, the dict _doesn't_ force the user to be explicit. 

Named tuples are interesting, although, it is something we could upgrade to in the future after a plain tuple interface if it's something we thought was valuable.

```python
for row in iter_rows_namedtuple():
  foo = row.Foo
  bar = row.Bar
  ...
```

Although, this still has the problem that it allows the user to lazily get much more data than they actually want to use (ie, with implicitly unused columns).

Reply:
1. the method accepts a list of column names
2. the common use patterns of dict need a key to read the value
3. even with tuples, a user can easily make the mistake of loading more data than they want and use the Python unpacking syntax to only read the interested columns.

Reply:
> even with tuples, a user can easily make the mistake of loading more data than they want and use the Python unpacking syntax to only read the interested columns.

Can you describe this a bit more? What is the unpackaging syntax that would let them load more data than they technically need?

```python
for foo, bar, baz in iter_rows_tuple():
  do_something(foo, bar)
  # user doesn't use baz
  ...
```

^ I consider this case to be "okay", as they still had to be explicit, and hopefully they'll have a linter warning them that they have an unused variable.

Reply:
Not when you do:
```python
for a, b, c, *_ in table.iter_rows():
    ...
```

Reply:
Sure; but in this case, they are really opting into it.

Reply:
I did some looking into this to determine what precedent is.  I think the most relevant component of the Python standard library is `csv`, which you can see at https://docs.python.org/3/library/csv.html.  In `csv`, it was determined that they needed to support both reading rows as tuples AND reading rows as dicts.  Their naming is inconsistent, so they clearly added the methods at different times.

After going through this exercise, I'm strongly of the view that we should support both tuples and dicts because they are both common access patterns that users want.  Namedtuple is something that we could add in the future if there is demand.

As far as the safety of tuple vs dict goes, there is a tradeoff.  tuple is resistant to changes in the number of columns and renamings, but it is susceptible to reorderings.  dict is resistant to extra columns and reordering, but it is susceptible to renamings.  We should provide both access patterns so that users can decide which downsides they want to accept.

This code will run on the demo system if you want to play with it.
```python
fn = "/data/Iris/iris.csv"

import csv

# read as a tuple
with open(fn) as csv_file:
    csv_reader = csv.reader(csv_file, delimiter=',')
    for row in csv_reader:
        print(row)

# read as a tuple and expand into variables
with open(fn) as csv_file:
    csv_reader = csv.reader(csv_file, delimiter=',')
    for seplen, sepwidth, petallen, petalwidth, kind in csv_reader:
        print(seplen, sepwidth, petallen, petalwidth, kind)

# read as a dict
with open(fn) as csv_file:
    csv_reader = csv.DictReader(csv_file)
    for row in csv_reader:
        print(row)
 ```

Reply:
I've done some more looking at this topic.  We may only want to support `namedtuple`.  It does everything, and it has better perf characteristics than dict.  **we need to verify that any DH column name will work as a `namedtuple` field.  Mappings aren't a great UX for this case.**

1. `namedtuple` supports positional access.
2. `namedtuple` supports access by name.
3. `namedtuple` is reported to be higher performance with less overhead than `dict`.
4. Here is a discussion of performance:  https://stackoverflow.com/questions/2646157/what-is-the-fastest-to-access-struct-like-object-in-python

Here is an expanded example to look at the iteration.

```python
import csv
from collections import namedtuple

fn = "/data/Iris/iris.csv"
Iris = namedtuple('Iris', ["seplen", "sepwidth", "petallen", "petalwidth", "kind"])


# read as a tuple
with open(fn) as csv_file:
    csv_reader = csv.reader(csv_file, delimiter=',')
    for row in csv_reader:
        print(row)

# read as a tuple and expand into variables
with open(fn) as csv_file:
    csv_reader = csv.reader(csv_file, delimiter=',')
    for seplen, sepwidth, petallen, petalwidth, kind in csv_reader:
        print(seplen, sepwidth, petallen, petalwidth, kind)

# read as a dict
with open(fn) as csv_file:
    csv_reader = csv.DictReader(csv_file)
    for row in csv_reader:
        print(row)

# read as a namedtuple

def namedreader(f):
    csv_reader = csv.reader(csv_file, delimiter=',')
    for row in csv_reader:
        yield Iris(*row)

with open(fn) as csv_file:
    csv_reader = namedreader(csv_file)
    for row in csv_reader:
        print(row)

with open(fn) as csv_file:
    csv_reader = namedreader(csv_file)
    for seplen, sepwidth, petallen, petalwidth, kind in csv_reader:
        print(seplen, sepwidth, petallen, petalwidth, kind)

with open(fn) as csv_file:
    csv_reader = namedreader(csv_file)
    for row in csv_reader:
        print(row.seplen, row.sepwidth, row.petallen, row.petalwidth, row.kind)
```

Reply:
I think offering `iter_rows_dict`, `iter_chunks_dict`, `iter_rows_tuple`, and `iter_chunks_tuple` is good. We may follow what Pandas does, just `iter_rows`, `iter_chunks` for dict output, but suffix `tuples` for tuple output.
@chipkent @devinrsmith your thoughts?